src/mutex.rs

```diff
@@ -0,0 +1,91 @@
+//! Implementation of a critical section based mutex that also implements the `mutex-trait`.
+//!
+//! ## Safety
+//!
+//! Note that this is only safe in single core applications.
+
+use core::cell::UnsafeCell;
+
+pub extern crate mutex_trait;
+pub use mutex_trait::Mutex;
+
+/// A spinlock and critical section section based mutex.
+pub struct CriticalSectionSpinLockMutex<T> {
+    data: spin::Mutex<T>,
+}
+
+impl<T> CriticalSectionSpinLockMutex<T> {
+    /// Create a new mutex
+    pub const fn new(data: T) -> Self {
+        CriticalSectionSpinLockMutex {
+            data: spin::Mutex::new(data),
+        }
+    }
+}
+
+impl<T> mutex_trait::Mutex for &'_ CriticalSectionSpinLockMutex<T> {
+    type Data = T;
+
+    fn lock<R>(&mut self, f: impl FnOnce(&mut Self::Data) -> R) -> R {
```

Comment:
In what scenario would this Mutex be used? If it's used to share a resource on the same core, its an instant deadlock right? In which case you'd only want to use it cross core, what benefits does blocking interrupts on the other core have?

Reply:
This one is the replacement for the interrupt::free with a mutex inside. I use this construction where I need to share access to a peripheral etc.

It prevents interference by an interrupt handler by the outer interrupt::free and from another core by the mutex.

The outer interrupt::free will prevent it from deadlocking by an interrupt as the interrupts are disabled. 

If indeed used recursively it would deadlock. (I think this is prevented by the weird &mut self syntax by the borrow checker).